exe/Main.hs

```diff
@@ -1,21 +1,61 @@
 -- Copyright (c) 2019 The DAML Authors. All rights reserved.
 -- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE NamedFieldPuns    #-}
 {-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE RecordWildCards   #-}
 module Main(main) where
 
-import           Ide.Arguments (Arguments (..), GhcideArguments (..),
-                                getArguments)
-import           Ide.Main      (defaultMain)
-import           Plugins
+import           Data.Function                ((&))
+import           Development.IDE.Types.Logger (Priority (Debug, Info),
+                                               WithPriority (WithPriority, priority),
+                                               cfilter, cmap,
+                                               withDefaultRecorder)
+import qualified Development.IDE.Types.Logger as Logger
+import           Ide.Arguments                (Arguments (..),
+                                               GhcideArguments (..),
+                                               getArguments)
+import           Ide.Main                     (defaultMain)
+import qualified Ide.Main                     as IdeMain
+import qualified Plugins
+import           Prettyprinter                (Doc, Pretty (pretty))
+import qualified System.Log                   as HsLogger
+
+data Log
+  = LogIdeMain IdeMain.Log
+  | LogPlugins Plugins.Log
+  deriving Show
+
+instance Pretty Log where
+  pretty log = case log of
+    LogIdeMain ideMainLog -> pretty ideMainLog
+    LogPlugins pluginsLog -> pretty pluginsLog
```

Comment:
I think my main reservation about this entire PR is really the boilerplate. Each function that wants logging needs a logger passed, and yeah the delegation that each Pretty instance and `logToPriority` function does is a bit annoying. This annoyance might extend to future `logToX`things. I'm just worried that for a contributor wants logging that this might be too annoying.

The above particular example is extra bad because I didn't do what I started doing later which was not to name the log variable too specifically.

Using TH or generics to reduce boilerplate I don't mind, if there's a clear explanation for it probably in the the contributor guide, and the generic function itself. I was planning on writing an explanation for this stuff anyway so maybe it's worth it.

Reply:
I've been thinking about dealing with the priority issue like this:
```
data WithSeverity a = WithSeverity { getMsg :: a, getSeverity :: Severity }

data Log1 = ...
data Log2 = ALog1 Log1 | ...

adapt :: LogAction m (WithSeverity Log2) -> LogAction m (WithSeverity Log1)
adapt = cmap (fmap ALog1)

logWithSeverity :: LogAction m (WithSeverity msg) -> Severity -> msg -> m ()
```

Then you'd end up stating the severities dynamically at the log site, but TBH I think that's okay, given that most messages are logged in only one place. And people are used to seeing `logDebug` etc. at the log site.

Reply:
See also https://github.com/co-log/co-log-core/issues/13

Reply:
I was considering that before but it directly opposed part of https://github.com/haskell/haskell-language-server/pull/2558#discussion_r777182269
> The priority is subjective and decided by the application, not by the `logMsg` call site.

Priorities being separate pros:
- more centralized so glancing over all priorities is easier
- some things don't have natural priorities (yet they would still have to add a pattern matching branch or use catchall branch)
- every log constructor is prioritized the same way (could also be a Con)
- any priority logic involving the logged values can be written here
- less noise for code reader (usually priorities are noise?)

Priorities being separate cons:
- more boilerplate
  - function(s) has to exist somewhere to do the conversion
  - pattern matching syntax requires writing the constructor (except for catch-all)
  - a significant amount of pattern matching branches delegate to some other `logToPriority` function

I think there was exactly 1 time so far where there was some priority logic that depended on the logged values and it felt cool to have that logic separate from the main code.

My reservations are totally about whether other people think it's tolerable to have these extra steps to log stuff. The boilerplate doesn't bother me much personally...

Reply:
Hmmm. I can see why you're doing that, but I'm not 100% sure it's desirable. I guess I have a sort of status-quo argument for not doing it: the old system set priorities at the log site; most other logging systems in other languages (that I've used) set priorities at the log site; it can't be *that* bad.

I'm also not sure that it makes it easier to manipulate priorities. It depends where that happens. How would I do something like the following?

```
suppressErrors  :: LogAction m (WithSeverity msg) -> LogAction m (WithSeverity m)
suppressErrors = cmap (\(WithSeverity sev msg) -> WithSeverity (if sev == Error then Warning else sev) msg)

-- Suppress errors from this subsystem, they're not fatal for us
doSomething (suppressErrors log)
```

It might also make the work simpler. Step 1 could leave priorities at the log sites as now, and then step 2 could pull them out. It might make this change easier to land?

Reply:
Yeah I've been thinking that logs and filtering seem pretty intrinsically tied. I guess it doesn't matter what your backend is, you always want a way to be able to filter messages. You might be able to filter them based on the their constructor and values, but we have no way of passing that information to down stream, and so having a value that represents that information is good.

I don't think there is a way of doing your example with `logToPriority` functions since all of the priorities get finally calculated at the end.

Ironically I'd have to change a lot of things in the PR to put priorities back at the log sites, but I don't mind doing that. Actually, I'll just do that since if anybody objects I can just revert to the current state which I think is ready for review.

Reply:
Sorry for suggesting such an annoying change! :sweat_smile: